server/src/main/java/com/microsoft/java/bs/core/internal/reporter/TestReportReporter.java

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package com.microsoft.java.bs.core.internal.reporter;
+
+import ch.epfl.scala.bsp4j.TaskId;
+import ch.epfl.scala.bsp4j.TaskStartParams;
+import ch.epfl.scala.bsp4j.TestStatus;
+import ch.epfl.scala.bsp4j.extended.TestFinishEx;
+import ch.epfl.scala.bsp4j.extended.TestStartEx;
+
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationResult;
+import org.gradle.tooling.events.ProgressEvent;
+import org.gradle.tooling.events.StartEvent;
+import org.gradle.tooling.events.test.JvmTestOperationDescriptor;
+import org.gradle.tooling.events.test.TestFailureResult;
+import org.gradle.tooling.events.test.TestSkippedResult;
+import org.gradle.tooling.events.test.TestSuccessResult;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import ch.epfl.scala.bsp4j.StatusCode;
+import ch.epfl.scala.bsp4j.TaskFinishParams;
+import ch.epfl.scala.bsp4j.TestReport;
+import org.gradle.tooling.internal.consumer.DefaultTestAssertionFailure;
+
+/**
+ * Implements {@link ProgressReporter} to record test results.
+ * Test summary report. e.g. number of fails.
+ */
+public class TestReportReporter extends ProgressReporter {
+
+  private final BuildTargetIdentifier btId;
+  private int successCount;
+  private int skippedCount;
+  private int failureCount;
+  private String exception;
+  private long testDuration;
+
+  /**
+   * initialise.
+   *
+   * @param btId the build target being tested.
+   */
+  public TestReportReporter(BuildTargetIdentifier btId, BuildClient client, String originId) {
+    super(client, originId);
+    this.btId = btId;
+    successCount = 0;
+    skippedCount = 0;
+    failureCount = 0;
+    exception = null;
+    testDuration = 0;
+  }
+
+  @Override
+  public void statusChanged(ProgressEvent event) {
+    if (client != null) {
+      if (event.getDescriptor() instanceof JvmTestOperationDescriptor descriptor) {
+        // only report on methods
+        if (descriptor.getClassName() != null && descriptor.getMethodName() != null) {
```

Comment:
I found that only report progress when class name and method name are not null might have some issue.

Considering:

```java
@TestFactory
  Stream<DynamicContainer> testContainers() {
    return Stream.of(dynamicContainer("Container", List.of(
      dynamicTest("Test", () -> {assertTrue(false);})
    )));
  }
```
The result hierarchy from IDEA is:

![image](https://github.com/microsoft/build-server-for-gradle/assets/6193897/d8a6b4cd-d6fe-45a1-a87a-16d979af4781)

Corresponds to the output of gradle tooling api：

 ```
PassingTests                     // 'PassingTests started' [className: "PassingTests"] [methodName: null]
    |-- testContainers            // 'Test suite 'testContainers() started' [className: null] [methodName: null]
             |-- Container        // 'Container started' [className: null] [methodName: null]
                      |-- Test    // 'Test started' [className: "PassingTests"] [methodName: "testContainers()[1][1]"]
```

I'm still thinking that how to display this hierarchy via BSP. It looks to me that we need to take the `name` and `parent` field of the `descriptor` into consideration as well.

Reply:
A better example:

```
@TestFactory
Collection<DynamicContainer> testContainer() {
    List<DynamicContainer> containers = new ArrayList<>();
    containers.add(
            dynamicContainer("First Container", List.of(
                    dynamicTest("First test of first container", () -> {
                    }),
                    dynamicTest("Second test of first container", () -> {
                    })
            ))
    );
    containers.add(
            dynamicContainer("Second Container", List.of(
                    dynamicTest("First test of second container", () -> {
                    }),
                    dynamicTest("Second test of second container", () -> {
                    })
            ))
    );
    return containers;
}
```

Which should be able to rendered like this in VS Code:

![image](https://github.com/microsoft/build-server-for-gradle/assets/6193897/2abb6ddf-4901-45be-92be-a46af5e6ffe1)

Reply:
I've got a feeling that the introduced new fields: `className` and `methodName` may not be suitable to present the test progress/result for the above example.

To summarize the current properties that we can use:

1. The client needs to know whether this test event is a start or finish. this can be achieved via the data kind: `test-start` / `test-finish`
2. For test finish event, the client needs to know the result of it. This can be achieved by the `TestStatus` in the data.
3. For failed cases, field `stacktrace` is needed. (Can we use the field `message` to store the stacktrace?)
4. Duration of each test case is needed, this can be calculated by checking the `eventTime`
5. Then, for each test case, we need to use `displayName` to store the content displayed in the ui. Take the above case as example: `First Container` or `First test of first container`.
6. We need another field acts asthe `test id` to let the client target the current test case, at the mean time, the field should be able to contain the hierarchy information for the test execution. Still taking the above example: use the `JvmTestOperationDescriptor.getName()` to get the current test name, and use `/` to present different level of the hierarchy. The hierarchy can be restored via `JvmTestOperationDescriptor.getParent()`. So, during the execution, we will get following ids:

`com.example.project.PassingTests`
`com.example.project.PassingTests/testContainer()`
`com.example.project.PassingTests/testContainer()/testContainer()[1]`
`com.example.project.PassingTests/testContainer()/testContainer()[1]/testContainer()[1][1]`
...

The question is where should we store the ids, using `taskId` or introduce another field?

WDYT?

Reply:
I've added the test hierarchy into `taskId`.  I think it's fine for testing - to make sure you've got everything for reporting but I imagine someone could create a dynamic test with a `/` in the name which would cause issues with your parsing.
Might be better to have a `TestNameEx` type like...
```
class TestNameEx {
  String name;
  TestNameEx parent;
}
```
I'm still only reporting on test tasks where `methodName != null`.  Is that OK? Or do you need `TaskStart` and `TaskFinish` for suites and classes as well?

Reply:
Since `TaskId` has a member `parent`. How about using that to present the hierarchy? Then we can avoid `a dynamic test with a / in the name` problem.

Reply:
> I'm still only reporting on test tasks where methodName != null. Is that OK? Or do you need TaskStart and TaskFinish for suites and classes as well?

I prefer having start and finish for suite and classes as well, which is align with the test execution in gradle.

Reply:
> Since `TaskId` has a member `parent`. How about using that to present the hierarchy? Then we can avoid `a dynamic test with a / in the name` problem.

`TaskId#parent` is a `List<String>` not another `TaskId` so you can't chain them.  I guess we could add encode the tests and put them in the `parent` as a list like...
```
TestName1:com.example.project.PassingTests
TestName2:testContainer()
TestName3:testContainer()[1]
TestName4:testContainer()[1][1]
550e8400-e29b-41d4-a716-446655440000   // <-- real parent task id
```

It would be a bit messy though.

I could put a hierarchy into `TestStartEx` and `TestFinishEx`?

Reply:
How will the hierarchy look like in `TestStartEx`/`TestFinishEx`?

Reply:
I've added `TestName` in a new commit.  See what you think, we can always roll it back.
`TaskId` no longer has the encoded test hierarchy.
`TestStartEx` and `TestFinishEx` no longer have `SuiteName`, `ClassName`, `MethodName`.
`TestName` is hierarchical and defined as...
```java
  String displayName;
  String suiteName;
  String className;
  String methodName;
  TestName parent;
```
So now you can pick and choose what to display in VSCode at each level in the hierarchy.  `displayName` seems good for most levels.

You'll also get additional `TestStartEx` and `TestFinishEx` messages as class level and sometimes suite level, depending on whether Gradle reports it for that class.

The test pass, skip, fail counts haven't changed as they still only count the methods run.